extends TileMap
class_name GameTilemap

signal tilemap_completed

var random_spawn_floor_positions: Array # vsi še ne zasedeni tileti, kamor se lahko potem random spawna (floor - stray - no-stray - player)
var floor_global_positions: Array # original tileti tal

var stray_global_positions: Array
var no_stray_global_positions: Array
var wall_stray_global_positions: Array
var player_global_positions: Array 

var wall_tile_id: int = 3
var edge_tile_id: int = 1
var floor_tile_id: int = 0
var spawn_stray_tile_id: int = 5 # za home btne

onready var camera_position_node: Position2D = $CameraPosition
onready var background_room: TextureRect = $Background/Room


func _ready() -> void:
	
	add_to_group(Global.group_tilemap) # za scrolling in patterns
	Global.current_tilemap = self

	# set_color_theme
	get_tileset().tile_set_modulate(wall_tile_id, Global.color_wall)
	get_tileset().tile_set_modulate(edge_tile_id, Global.color_edge)
	get_tileset().tile_set_modulate(floor_tile_id, Global.color_floor)


func get_tiles():
	
	if Global.game_manager.game_data["game"] == Profiles.Games.ENIGMA:
		if Global.game_manager.game_settings["solutions_mode"]:
#			$SolutionLine.modulate = Color("#0a0a0a")
			$SolutionLine.modulate = Global.color_almost_black
		else:
			$SolutionLine.hide()
	
	# prečesi vse celice in določi globalne pozicije
	for x in get_used_rect().size.x: # širina v celicah
		for y in get_used_rect().size.y: # višina širina v celicah	 
			
			# pretvorba v globalno pozicijo
			var cell: Vector2 = Vector2(x, y) # grid pozicija celice
			var cell_local_position: Vector2 = map_to_world(cell)
			var cell_global_position: Vector2 = to_global(cell_local_position) # pozicija je levo-zgornji vogal celice
			
			# zaznavanje tiletov
			var cell_index = get_cellv(cell)
			match cell_index:
				0: # floor
					floor_global_positions.append(cell_global_position)
					random_spawn_floor_positions.append(cell_global_position)
				5: # stray spawn positions
					stray_global_positions.append(cell_global_position)
					set_cellv(cell, 0) # menjam za celico tal
					floor_global_positions.append(cell_global_position)
				2: # no stray
					no_stray_global_positions.append(cell_global_position)
					set_cellv(cell, 0)
					floor_global_positions.append(cell_global_position)
				4: # player 1 spawn position
					player_global_positions.append(cell_global_position)
					set_cellv(cell, 0)
					floor_global_positions.append(cell_global_position)
				6: # player 2 spawn position
					player_global_positions.append(cell_global_position)
					set_cellv(cell, 0)
					floor_global_positions.append(cell_global_position)
				8: # spawn wall stray
					wall_stray_global_positions.append(cell_global_position)
					set_cellv(cell, 0)
					floor_global_positions.append(cell_global_position)
	
	# pošljem v GM
	emit_signal("tilemap_completed", random_spawn_floor_positions, stray_global_positions, no_stray_global_positions, player_global_positions, wall_stray_global_positions)
	
	
func get_collision_tile_id(collider: Node2D, direction: Vector2): # collider je node ki se zaleteva in ne collision object
	
	# pozicija celice stene
	var collider_position = collider.position
	var colliding_cell_position = collider_position + direction * cell_size.x # dodamo celico v smeri gibanja, da ne izber pozicije pixla
	
	# index tileta
	var colliding_cell_grid_position: Vector2 = world_to_map(colliding_cell_position) # katera celica je bila zadeta glede na global coords
	var tile_index: int = get_cellv(colliding_cell_grid_position) # index zadete celice na poziciji v grid koordinatah
	
	return tile_index
